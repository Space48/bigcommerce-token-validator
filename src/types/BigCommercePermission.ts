
export enum BigCommercePermission {
    All = "all",
    Content = "store_v2_content",
    ContentRead = "store_v2_content_read_only",
    CheckoutContent = "store_content_checkout",
    CheckoutContentRead = "store_content_checkout_read_only",
    Customers = "store_v2_customers",
    CustomersRead = "store_v2_customers_read_only",
    CustomersLogin = "store_v2_customers_login",
    InformationSettings = "store_v2_information",
    InformationSettingsRead = "store_v2_information_read_only",
    Marketing = "store_v2_marketing",
    MarketingRead = "store_v2_marketing_read_only",
    Orders = "store_v2_orders",
    OrdersRead = "store_v2_orders_read_only",
    OrderTransactions = "store_v2_order_transactions",
    OrderTransactionsRead = "store_v2_order_transactions_read_only",
    CreatePayments = "store_payments_access_token_create",
    GetPaymentMethodsRead = "store_payments_methods_read",
    Products = "store_v2_products",
    ProductsRead = "store_v2_products_read_only",
    Themes = "store_themes_manage",
    ThemesRead = "store_themes_read_only",
    Carts = "store_cart",
    CartsRead = "store_cart_read_only",
    Checkouts = "store_checkout",
    CheckoutsRead = "store_checkout_read_only",
    SitesRoutes = "store_sites",
    SitesRoutesRead = "store_sites_read_only",
    ChannelSettingsd = "store_channel_settings",
    ChannelSettingsRead = "store_channel_settings_read_only",
    ChannelListings = "store_channel_listings",
    ChannelListingsRead = "store_channel_listings_read_only",
    StorefrontApiTokens = "store_storefront_api",
    StorefrontApiCustomerImpersonationTokens = "store_storefront_api_customer_impersonation"
}
