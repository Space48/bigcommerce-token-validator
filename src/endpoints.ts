import { BigCommercePermission } from "./types/BigCommercePermission"


export interface BigCommercePermissionRequest {
    enabled?: boolean;
    method: "GET" | "POST" | "PUT" | "DELETE";
    version: "v2" | "v3";
    path: string;
    test?: (response: any) => boolean;
}
export type BigCommercePermissionEndpointCollection = Partial<{
    [key in BigCommercePermission]: BigCommercePermissionRequest;
}>;

export const BigCommercePermissionEndpoints: BigCommercePermissionEndpointCollection = {
    [BigCommercePermission.Content]: {
        method: "POST",
        version: "v2",
        path: "/pages"
    },
    [BigCommercePermission.ContentRead]: {
        method: "GET",
        version: "v2",
        path: "/pages"
    },
    [BigCommercePermission.CheckoutContent]: {
        method: "POST",
        version: "v3",
        path: "/content/scripts"
    },
    [BigCommercePermission.CheckoutContentRead]: {
        method: "GET",
        version: "v3",
        path: "/content/scripts/1"
    },
    [BigCommercePermission.Customers]: {
        method: "POST",
        version: "v3",
        path: "/customers"
    },
    [BigCommercePermission.CustomersRead]: {
        method: "GET",
        version: "v3",
        path: "/customers"
    },
    // [BigCommercePermission.CustomersLogin]: { // TODO: Add case for this as not mgmt api
    //     enabled: false,
    //     method: "GET",
    //     version: "v2",
    //     path: "/*"
    // },
    // [BigCommercePermission.InformationSettings]: { // No methods to test against
    //     enabled: false,
    //     method: "GET",
    //     version: "v2",
    //     path: "/*"
    // },
    [BigCommercePermission.InformationSettingsRead]: {
        method: "GET",
        version: "v2",
        path: "/time",
        test: (response: any) => {
            return Boolean(response?.body?.time)
        },
    },
    [BigCommercePermission.Marketing]: {
        method: "POST",
        version: "v2",
        path: "/banners"
    },
    [BigCommercePermission.MarketingRead]: {
        method: "GET",
        version: "v2",
        path: "/banners"
    },
    [BigCommercePermission.Orders]: {
        method: "POST",
        version: "v2",
        path: "/orders"
    },
    [BigCommercePermission.OrdersRead]: {
        method: "GET",
        version: "v2",
        path: "/orders"
    },
    // [BigCommercePermission.OrderTransactions]: { // No methods to test against
    //     method: "POST",
    //     version: "v3",
    //     path: "/orders/1/transactions"
    // },
    [BigCommercePermission.OrderTransactionsRead]: {
        method: "GET",
        version: "v3",
        path: "/orders/1/transactions"
    },
    [BigCommercePermission.CreatePayments]: {
        method: "POST",
        version: "v3",
        path: "/payments/access_tokens"
    },
    [BigCommercePermission.GetPaymentMethodsRead]: {
        method: "GET",
        version: "v2",
        path: "/payments/methods"
    },
    [BigCommercePermission.Products]: {
        method: "POST",
        version: "v3",
        path: "/catalog/products"
    },
    [BigCommercePermission.ProductsRead]: {
        method: "GET",
        version: "v3",
        path: "/catalog/products"
    },
    [BigCommercePermission.Themes]: {
        method: "POST",
        version: "v3",
        path: "/themes"
    },
    [BigCommercePermission.ThemesRead]: {
        method: "GET",
        version: "v3",
        path: "/themes"
    },
    [BigCommercePermission.Carts]: {
        method: "POST",
        version: "v3",
        path: "/carts"
    },
    [BigCommercePermission.CartsRead]: {
        method: "GET",
        version: "v3",
        path: "/carts/1"
    },
    [BigCommercePermission.Checkouts]: {
        method: "POST",
        version: "v3",
        path: "/checkouts/1/billing-address"
    },
    [BigCommercePermission.CheckoutsRead]: {
        method: "GET",
        version: "v3",
        path: "/checkouts/1"
    },
    [BigCommercePermission.SitesRoutes]: {
        enabled: false,
        method: "GET",
        version: "v2",
        path: "/*"
    },
    [BigCommercePermission.SitesRoutesRead]: {
        method: "GET",
        version: "v3",
        path: "/sites/1/routes"
    },
    [BigCommercePermission.ChannelSettingsd]: {
        method: "POST",
        version: "v3",
        path: "/channels/currency-assignments"
    },
    [BigCommercePermission.ChannelSettingsRead]: {
        method: "GET",
        version: "v3",
        path: "/channels/currency-assignments"
    },
    [BigCommercePermission.ChannelListings]: {
        method: "POST",
        version: "v3",
        path: "/channels/1/listings"
    },
    [BigCommercePermission.ChannelListingsRead]: {
        method: "GET",
        version: "v3",
        path: "/channels/1/listings"
    },
    [BigCommercePermission.StorefrontApiTokens]: {
        enabled: false,
        method: "GET",
        version: "v2",
        path: "/*"
    },
    [BigCommercePermission.StorefrontApiCustomerImpersonationTokens]: {
        enabled: false,
        method: "GET",
        version: "v2",
        path: "/*"
    }
}