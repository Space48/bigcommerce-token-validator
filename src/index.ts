import BigCommerce from 'node-bigcommerce';
import { BigCommercePermission } from './types/BigCommercePermission';
import { BigCommercePermissionEndpoints } from './endpoints';



const getClient = (creds: {
    clientId: string,
    accessToken: string,
    storeHash: string,

}, version: "v2" | "v3") => {
    return new BigCommerce({
        logLevel: 'info',
        responseType: 'json',
        headers: { 'Accept-Encoding': '*' }, // Override headers (Overriding the default encoding of GZipped is useful in development)
        apiVersion: version, // Default is v2
        ...creds
    });
}

const isUnauthorized = (status: number) => (status === 403)
const isNotFound = (status: number) => (status === 404)
const isNotError = (status: number) => (status >= 200 && status < 300)
const formatError = (status: number) => (status === 422) || (status === 400);
const isSuccessful = (status: number) => !isUnauthorized(status) && (isNotError(status) || isNotFound(status) || formatError(status));

const processPermissions = (permissions: BigCommercePermission[]): BigCommercePermission[] => {
    if(permissions.some(p => p === BigCommercePermission.All)) {
        return Object.keys(BigCommercePermissionEndpoints)
            .map(p => p as BigCommercePermission)
            .filter(p => BigCommercePermissionEndpoints[p] !== undefined)
            .filter(p => BigCommercePermissionEndpoints[p]!.enabled !== false)
            .filter(p => p !== BigCommercePermission.All)
    }
    return permissions
    .filter(p => BigCommercePermissionEndpoints[p] !== undefined)
    .filter(p => BigCommercePermissionEndpoints[p]!.enabled !== false);
}

export const DoAudit = async (creds: {
    clientId: string,
    accessToken: string,
    storeHash: string,

}, permissions: BigCommercePermission[]) => {
    const results = processPermissions(permissions).map(async (permission) => {
        const endpoint = BigCommercePermissionEndpoints[permission]!;
        if (endpoint.enabled === false) {
            console.log(`Permission ${permission} is disabled`);
            return { permission, result: false };
        }
        const client = getClient(creds, endpoint.version);

        let response = undefined;

        if (endpoint.method === "POST") {
            response = await client.post(endpoint.path, {  })
            .then((r: any) => ({ status: 200, body: r }))
            .catch((e: Error & {code: number, responseBody: string}) => ({ status: e.code, body: e.responseBody }));
        } else if (endpoint.method === "GET") {
            response = await client.get(endpoint.path)
            .then((r: any) => ({ status: 200, body: r }))
            .catch((e: Error & {code: number, responseBody: string}) => ({ status: e.code, body: e.responseBody }));
        } else {
            console.log(`Permission ${permission} has an invalid method`);
            return { permission, result: false };
        }
        const isSuccessfulResult = endpoint.test ? endpoint.test(response) : isSuccessful(response.status);
        return { permission, result: isSuccessfulResult, status: response.status, message: !isSuccessfulResult ? response.body: undefined };
    });
    return Promise.all(results);

}