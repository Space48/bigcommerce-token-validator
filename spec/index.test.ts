require('dotenv').config()

import { DoAudit } from '../src/index';
import { BigCommercePermission } from '../src/types/BigCommercePermission';

const noPermsCreds = {
    // Get from env var
    clientId: process.env.BC_CREDS_LOGGING_ONLY_CLIENTID!,
    accessToken: process.env.BC_CREDS_LOGGING_ONLY_ACCESSTOKEN!,
    storeHash: process.env.BC_CREDS_LOGGING_ONLY_STOREHASH!,
}

const allButCustomersWriteCreds = {
    clientId: process.env.BC_CREDS_ALL_BUT_CUSTOMERS_WRITE_CLIENTID!,
    accessToken: process.env.BC_CREDS_ALL_BUT_CUSTOMERS_WRITE_ACCESSTOKEN!,
    storeHash: process.env.BC_CREDS_ALL_BUT_CUSTOMERS_WRITE_STOREHASH!,
};

describe('index', () => {

    describe('None except logs', () => {

        it('should audit', async () => {
            const permissions = [BigCommercePermission.All];

            const results = await DoAudit(noPermsCreds, permissions);
            expect(results).toMatchSnapshot();
        });
    });

    describe('All except customer write', () => {
        it('should audit all', async () => {
            const permissions = [BigCommercePermission.All];
            const results = await DoAudit(allButCustomersWriteCreds, permissions);
            expect(results).toMatchSnapshot();

            const customerResult = results.find(r => r.permission === BigCommercePermission.Customers);
            const customerReadResult = results.find(r => r.permission === BigCommercePermission.CustomersRead);

            expect(customerResult).toBeDefined();
            expect(customerReadResult).toBeDefined();

            expect(customerResult?.result).toEqual(false);
            expect(customerReadResult?.result).toEqual(true);

            const resultsWithoutCustomerWrite = results.filter(r => r.permission !== BigCommercePermission.Customers);
            expect(resultsWithoutCustomerWrite.map(r => r.result).every(r => r === true)).toBeTruthy();

            return;

        });
    });
});