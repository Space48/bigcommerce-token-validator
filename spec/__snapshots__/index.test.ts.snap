// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index All except customer write should audit all 1`] = `
[
  {
    "message": undefined,
    "permission": "store_v2_content",
    "result": true,
    "status": 400,
  },
  {
    "message": undefined,
    "permission": "store_v2_content_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_content_checkout",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_content_checkout_read_only",
    "result": true,
    "status": 422,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_customers",
    "result": false,
    "status": 403,
  },
  {
    "message": undefined,
    "permission": "store_v2_customers_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_v2_information_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_v2_marketing",
    "result": true,
    "status": 400,
  },
  {
    "message": undefined,
    "permission": "store_v2_marketing_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_v2_orders",
    "result": true,
    "status": 400,
  },
  {
    "message": undefined,
    "permission": "store_v2_orders_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_v2_order_transactions_read_only",
    "result": true,
    "status": 404,
  },
  {
    "message": undefined,
    "permission": "store_payments_access_token_create",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_payments_methods_read",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_v2_products",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_v2_products_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_themes_manage",
    "result": true,
    "status": 400,
  },
  {
    "message": undefined,
    "permission": "store_themes_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_cart",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_cart_read_only",
    "result": true,
    "status": 404,
  },
  {
    "message": undefined,
    "permission": "store_checkout",
    "result": true,
    "status": 404,
  },
  {
    "message": undefined,
    "permission": "store_checkout_read_only",
    "result": true,
    "status": 404,
  },
  {
    "message": undefined,
    "permission": "store_sites_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_channel_settings",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_channel_settings_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": undefined,
    "permission": "store_channel_listings",
    "result": true,
    "status": 422,
  },
  {
    "message": undefined,
    "permission": "store_channel_listings_read_only",
    "result": true,
    "status": 200,
  },
]
`;

exports[`index None except logs should audit 1`] = `
[
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_content",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_content_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_content_checkout",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_content_checkout_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_customers",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_customers_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": undefined,
    "permission": "store_v2_information_read_only",
    "result": true,
    "status": 200,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_marketing",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_marketing_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_orders",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_orders_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_order_transactions_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_payments_access_token_create",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_payments_methods_read",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_products",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_v2_products_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_themes_manage",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_themes_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_cart",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_cart_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_checkout",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_checkout_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_sites_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_channel_settings",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_channel_settings_read_only",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_channel_listings",
    "result": false,
    "status": 403,
  },
  {
    "message": "{"status":403,"title":"You don't have a required scope to access the endpoint","type":"https://developer.bigcommerce.com/api-docs/getting-started/api-status-codes","errors":{}}",
    "permission": "store_channel_listings_read_only",
    "result": false,
    "status": 403,
  },
]
`;
